keanu-project/src/main/java/io/improbable/keanu/algorithms/graphtraversal/DifferentiableChecker.java

```diff
@@ -0,0 +1,128 @@
+        Set<Vertex> allParents = new HashSet<>();
+        for (Vertex vertex : vertices) {
+            allParents.addAll(vertex.getParents());
+        }
+        return allParents;
+    }
+
+    private static boolean diffableOrConstantUptoNextRV(Collection<Vertex> vertices, Set<Vertex> constantValueVerticesCache) {
+        return bfs(vertices,
+            vertex -> isNonDiffableAndNotConstant(vertex, constantValueVerticesCache),
+            vertex -> !vertex.isProbabilistic(),
+            doNothing -> {
+            });
+    }
+
+    private static boolean bfs(Collection<Vertex> vertices,
+                               Predicate<Vertex> failureCondition,
+                               Predicate<Vertex> shouldAddParents,
+                               Consumer<Collection<Vertex>> successfullyVisitedConsumer) {
+
+        Queue<Vertex> queue = new ArrayDeque<>(vertices);
+        Set<Vertex> visited = new HashSet<>(vertices);
+
+        while (!queue.isEmpty()) {
+            Vertex visiting = queue.poll();
+
+            if (failureCondition.test(visiting)) {
+                return false;
+            }
+
+            if (shouldAddParents.test(visiting)) {
+                queueUnvisitedParents(visiting, queue, visited);
+            }
+        }
+
+        successfullyVisitedConsumer.accept(visited);
+        return true;
+    }
+
+    private static void queueUnvisitedParents(Vertex vertex, Queue<Vertex> queue, Set<Vertex> visited) {
+        Collection<Vertex> nextVertices = vertex.getParents();
+        for (Vertex next : nextVertices) {
+            if (!visited.contains(next)) {
+                queue.offer(next);
+                visited.add(next);
+            }
+        }
+    }
+
+    private static boolean isNonDiffableAndNotConstant(Vertex vertex, Set<Vertex> constantValueVerticesCache) {
+        return !vertex.isDifferentiable() &&
+            !isVertexValueConstant(vertex, constantValueVerticesCache);
+    }
+
+    private static boolean isVertexValueConstant(Vertex vertex, Set<Vertex> constantValueVerticesCache) {
+        if (isValueKnownToBeConstant(vertex, constantValueVerticesCache)) {
+            return true;
+        }
+
+        return bfs(Collections.singletonList(vertex),
+            DifferentiableChecker::isUnobservedProbabilistic,
+            visiting -> !isValueKnownToBeConstant(visiting, constantValueVerticesCache),
+            constantValueVerticesCache::addAll);
+    }
+
+    private static boolean isUnobservedProbabilistic(Vertex vertex) {
+        return vertex.isProbabilistic() && !vertex.isObserved();
+    }
+
+    // We know whether these are constant. For cases such as a MultiplicationVertex we would need to
+    // explore its parents to ensure its constant.
+    private static boolean isValueKnownToBeConstant(Vertex vertex, Set<Vertex> constantValueVerticesCache) {
+        return vertex instanceof ConstantVertex || vertex.isObserved() || constantValueVerticesCache.contains(vertex);
```

Comment:
the double meaning of "constant" here is yucky, but we have a task on the backlog to rename `ConstantVertex` which will make this better.